### ISSUE TITLE
Potential regression with generic descriptors (v1.1.314 -> v1.1.315)

### ISSUE THREAD

**Issue by runfalk** [microsoft/pyright#5455]
I don't know enough type lingo to describe it precisely, but in some circumstances where an unbound generic type parameter was accepted before it now raises an error. I don't know if this is limited to descriptors, but that's where I have been able to reproduce it.

**To Reproduce**
The following program passes type checking in `1.1.314` but fails in `1.1.315` and `1.1.316`. This code is a bit dumb, but it's a small example that exhibits the same behavior as the actual implementation. The real code is part of an abstraction on top of dataclasses for for an ORM style framework.

```python
import dataclasses
import typing as t

T = t.TypeVar("T", bound=int | float)

class ListAttr(t.Generic[T]):
    inner: list[T]

    def __init__(self, inner: list[T]) -> None:
        self.inner = inner

    def __get__(self, instance: object | None, owner: type[t.Any] | None = None) -> list[T]:
        if instance is None:
            raise AttributeError("No such attribute")
        return self.inner

    def __set__(self, instance: object, value: list[T]) -> None:
        self.inner = []


def default_factory(factory: t.Callable[[], list[T]]) -> ListAttr[T]:
    return ListAttr(factory())


class Container:
    not_working: ListAttr[int] = default_factory(list)
    working: ListAttr[int] = default_factory(lambda: [1])

c = Container()
print(c.not_working)  # []
print(c.working)  # [1]

t.reveal_type(c.not_working)  # list[int]
t.reveal_type(c.working)  # list[int]
```

**Expected behavior**
I expect this to pass like it did in `1.1.314` since the factory for an empty list is applicable for any `T`.

**Actual behavior**
In `1.1.315` and `1.1.316` I get this error.

```
/x.py
  /x.py:26:34 - error: Expression of type "ListAttr[T@ListAttr]" cannot be assigned to declared type "ListAttr[int]"
    "ListAttr[T@ListAttr]" is incompatible with "list[int]" (reportGeneralTypeIssues)
  /x.py:33:15 - information: Type of "c.not_working" is "list[int]"
  /x.py:34:15 - information: Type of "c.working" is "list[int]"
1 error, 0 warnings, 2 informations
```

**Additional context**
The workaround I have used for now is to introduce an optional parameter to `default_factory` that assigns a concrete type to `T`.

There seems to be something going on with descriptors specifically since the following code works:

```python
import dataclasses
import typing as t

T = t.TypeVar("T", bound=int | float)

class ListAttr(t.Generic[T]):
    inner: list[T]

    def __init__(self, inner: list[T]) -> None:
        self.inner = inner


def default_factory(factory: t.Callable[[], list[T]]) -> ListAttr[T]:
    return ListAttr(factory())


now_working: ListAttr[int] = default_factory(list)  # Works as expected
working: ListAttr[int] = default_factory(lambda: [1])
```

**Comment by erictraut** [comment]
Thanks for the bug report. The root cause of this problem was a long-standing bug that was previously masked. A recent change exposed it. I've fixed the root cause and added a test case to detect regressions in the future. This will be included in the next release of pyright.

**Comment by runfalk** [comment]
Thank you for fixing this!

If anyone runs into this in the meantime, a simpler fix is to pass an explicit type to the `list` constructor instead of introducing a new type parameter. Like so:

```python
class Container:
    workaround: ListAttr[int] = default_factory(list[int])  # works when type is explicitly specified
    working: ListAttr[int] = default_factory(lambda: [1])
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.318, which I just published. It will also be included in a future release of pylance.
